cli/parser/parser.go

```diff
@@ -186,86 +239,28 @@ func NewOptionDefinitionSet(optionDefinitions []*OptionDefinition, isStartupOpti
 // If args[start] corresponds to an option definition that is not known by the
 // option definition set, the returned values will be (nil, "", start+1). It is
 // up to the caller to decide how args[start] should be interpreted.
-func (s *OptionDefinitionSet) Next(args []string, start int) (optionDefinition *OptionDefinition, value string, next int, err error) {
+func (s *OptionDefinitionSet) Next(command string, args []string, start int) (optionDefinition *OptionDefinition, value string, next int, err error) {
 	if start > len(args) {
 		return nil, "", -1, fmt.Errorf("arg index %d out of bounds", start)
 	}
 	startToken := args[start]
-	var optValue *string
-	if strings.HasPrefix(startToken, "--") {
-		longName := strings.TrimPrefix(startToken, "--")
-		eqIndex := strings.Index(longName, "=")
-		if eqIndex >= 0 {
-			v := longName[eqIndex+1:]
-			longName = longName[:eqIndex]
-			optionDefinition = s.ByName[longName]
-			optValue = &v
-			if optionDefinition != nil && !optionDefinition.RequiresValue {
-				// Unlike command options, startup options don't allow specifying
-				// values for options that do not require values.
-				if s.IsStartupOptions {
-					return nil, "", -1, fmt.Errorf("in option %q: option %q does not take a value", startToken, optionDefinition.Name)
-				}
-				// Boolean options may specify values, but expansion options ignore
-				// values and output a warning. Since we canonicalize the options and
-				// remove the value ourselves, we should output the warning instead.
-				if !optionDefinition.HasNegative {
-					log.Warnf("option '--%s' is an expansion option. It does not accept values, and does not change its expansion based on the value provided. Value '%s' will be ignored.", optionDefinition.Name, *optValue)
-					optValue = nil
-				}
-			}
-		} else {
-			optionDefinition = s.ByName[longName]
-			// If the long name is unrecognized, check to see if it's actually
-			// specifying a bool flag like "noremote_upload_local_results"
-			if optionDefinition == nil && strings.HasPrefix(longName, "no") {
-				longName := strings.TrimPrefix(longName, "no")
-				optionDefinition = s.ByName[longName]
-				if optionDefinition != nil && !optionDefinition.HasNegative {
-					return nil, "", -1, fmt.Errorf("illegal use of 'no' prefix on non-boolean option: %s", startToken)
-				}
-				v := "0"
-				optValue = &v
-			}
-		}
-	} else if strings.HasPrefix(startToken, "-") {
-		shortName := strings.TrimPrefix(startToken, "-")
-		if !flagShortNamePattern.MatchString(shortName) {
-			return nil, "", -1, fmt.Errorf("invalid options syntax: %s", startToken)
-		}
-		optionDefinition = s.ByShortName[shortName]
+	option, needsValue, err := s.ParseOption(command, startToken)
+	if err != nil {
+		return nil, "", -1, err
 	}
-	if optionDefinition == nil {
+	if option == nil {
+		log.Printf("Unknown option %s for command %s", startToken, command)
```

Comment:
nit: remove this print, or move it to debug level?

Reply:
Yeah, this was supposed to be a debug statement; I'll demote it.